docs/msbuild/msbuild-batching.md

```diff
@@ -215,6 +215,116 @@ is not allowed.
+
+Metadata referencing in this case leads to batching - yielding possibly unexpected and unintended output:
+
+```output
+  i=[a/b.foo;c/d.foo;g/h.foo;g/h.foo]
+  i->MyPath=[;b.foo;b.foo;d.foo]
+```
+
+For each item instance the engine is applying metadata of all pre-existing item instances (that's why the `MyPath` is empty for the first item and contains `b.foo` for the second item), in case of more pre-existing instances this will lead to multiplication of the current item instance (that's why the `g/h.foo` item instance ocurring twice in the resulting list).
+
+To explicitly inform about this, possibly unintended, behavior newer version of MSBuild issue warning `MSB4120`:
+
+```output
+proj.proj(4,11):  warning MSB4120: Item 'i' definition within target is referencing self via metadata 'Filename' (qualified or unqualified). This can lead to unintended expansion and cross-applying of pre-existing items. More info: https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-batching#item-batching-on-self-referencing-metadata
+proj.proj(4,11):  warning MSB4120: Item 'i' definition within target is referencing self via metadata 'Extension' (qualified or unqualified). This can lead to unintended expansion and cross-applying of pre-existing items. More info: https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-batching#item-batching-on-self-referencing-metadata
+proj.proj(5,11):  warning MSB4120: Item 'i' definition within target is referencing self via metadata 'Filename' (qualified or unqualified). This can lead to unintended expansion and cross-applying of pre-existing items. More info: https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-batching#item-batching-on-self-referencing-metadata
+proj.proj(5,11):  warning MSB4120: Item 'i' definition within target is referencing self via metadata 'Extension' (qualified or unqualified). This can lead to unintended expansion and cross-applying of pre-existing items. More info: https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-batching#item-batching-on-self-referencing-metadata
+proj.proj(6,11):  warning MSB4120: Item 'i' definition within target is referencing self via metadata 'Filename' (qualified or unqualified). This can lead to unintended expansion and cross-applying of pre-existing items. More info: https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-batching#item-batching-on-self-referencing-metadata
+proj.proj(6,11):  warning MSB4120: Item 'i' definition within target is referencing self via metadata 'Extension' (qualified or unqualified). This can lead to unintended expansion and cross-applying of pre-existing items. More info: https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-batching#item-batching-on-self-referencing-metadata
+  i=[a/b.foo;c/d.foo;g/h.foo;g/h.foo]
+  i->MyPath=[;b.foo;b.foo;d.foo]
+```
+
+If the self-reference is intentional, you can use the [transform](../msbuild/msbuild-transforms.md) notation to avoid the warning:
```

Comment:
Correct. Transform is 1 for 1. In case of need for generating items (e.g. via leveraging wildcards) it's better to perform this via separate helper item or outside of the target context.

I'll update to make this less confusing and missleading